### PR TITLE
Add logic to export metrics for all certificates in the chain

### PR DESCRIPTION
**Pull request by Insannik**
Hello,

At Be-Mobile, we've been using cert-exporter for a few months to monitor thousands of certificates. We use it to get notified when critical certificates are about to expire. However, we found out the hard way that the current version of cert-exporter does not account for all certificates in the certificate chain which means we did not get notified when one of the intermediate certificates would expire.

To solve this issue, I've added an extra parameter/flag to the cert-exporter:
```
-include-full-cert-chain
      Set to true to expose metrics for all certificates in the chain (Default false).
```
Adding this flag when starting cert-exporter will export metrics for all certificates in the chain. Omitting the flag will result in cert-exporters original behaviour which means only the first certificate in the chain gets exported.

Feel free to let us know what you think.

Thanks!

### CODE REVIEW

**Review by joe-elliott — CHANGES_REQUESTED** [review]
Thank you for this submission and apologies that you had a incident that could have been avoided.

Two thoughts:
- Is there any reason we shouldn't just make this the default behavior instead of making it configurable?  Will this change existing metrics?
- Can we get a simple test that parses a file with multiple certs?

This is an excellent PR.  Thank you for taking the time and apologies for the oversight.

**Comment by Insannik** [comment]
Hi again,

First of all, don't worry as the incident wasn't extremely critical.

The changes wont change the existing metrics. The flag simply toggles between only exporting the first certificate in the chain and exporting all certificates in the chain. I didn't want to make this the default behaviour because I didn't want to change how cert-exporter currently works. When exporting the entire chain, you could get loads of metrics you don't actually need which, depending on the size of the environment, could lead to scrape timeouts. If you want this to become the default behaviour this is possible by simply defaulting the flag to true instead of false. At least this will still give the user the ability to toggle the functionality.

I'll check if I can write a small test to test parsing multiple certs in a file. I'll have to figure out how the current tests work in order to write one myself but I'll do my best.

**Comment by joe-elliott** [comment]
>  When exporting the entire chain, you could get loads of metrics you don't actually need which, depending on the size of the environment, could lead to scrape timeouts.

I think think this is unlikely and, to be honest, if their scrapes start timing out then that will get the operator to look closer and realize they had unmonitored certs.  Let's just make it happen by default, unconfigurable.  If it breaks something then we can add an option to turn it off.

> I'll have to figure out how the current tests work in order to write one myself but I'll do my best.
Tests are kind of gross b/c they're a stack of bash scripts, but hopefully it won't be too bad.  

Modify this to generate a file with multiple certs:
https://github.com/joe-elliott/cert-exporter/blob/master/test/files/genCerts.sh

File: .goreleaser.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ builds:
     goos:
       - linux
       - windows
+      - darwin
```
Comment (jenting) [inline .goreleaser.yml:10]: ❤️

File: .gitignore [hunk 2]
```diff
@@ -2,4 +2,5 @@
 dist/
 test/files/certs
 test/files/certsSibling
-test/files/kubeConfigSibling
\ No newline at end of file
+test/files/kubeConfigSibling
+*.sh
```
Comment (jenting) [inline .gitignore:6]: I think we should not ignore *.sh

**Comment by Insannik** [comment]
> > When exporting the entire chain, you could get loads of metrics you don't actually need which, depending on the size of the environment, could lead to scrape timeouts.
> 
> I think think this is unlikely and, to be honest, if their scrapes start timing out then that will get the operator to look closer and realize they had unmonitored certs. Let's just make it happen by default, unconfigurable. If it breaks something then we can add an option to turn it off.
> 
> > I'll have to figure out how the current tests work in order to write one myself but I'll do my best.
> > Tests are kind of gross b/c they're a stack of bash scripts, but hopefully it won't be too bad.
> 
> Modify this to generate a file with multiple certs:
> https://github.com/joe-elliott/cert-exporter/blob/master/test/files/genCerts.sh
> 
> Add a line here to look for the new metric:
> https://github.com/joe-elliott/cert-exporter/blob/master/test/files/test.sh#L57

Alright! I'll try and get this done asap.

**Comment by Insannik** [comment]
Hi again,

Sorry for taking so long. I finally got around to working on this PR some more.
So far I've changed the default behaviour to export metrics for all certificates in the chain instead of making it an option.

Next, I'll try and add some tests.

**Comment by joe-elliott** [comment]
No worries.  This is a good improvement and I'm patient :).

If you have issues with the tests let me know.  It doesn't have to be anything fancy.  Just add a couple certs to the same file and make sure their metrics exist.

**Comment by Insannik** [comment]
Hello again,

I've finally found some time to write/edit the tests.

I've edited 'genCerts.sh' to generate 2 extra certificates (CN: bundle-root and CN: example-bundle.be). Using 'cat' I put them both in the same file. Next, I changed test.sh to look for the 2 new entries using the ValidateMetrics method you wrote.

Feel free to check the changes thoroughly and let me know if I forgot anything.

**Review by joe-elliott — APPROVED** [review]
Fantastic fix.  Thank you!

You have probably saved some other SREs out there an unexpected expired cert.  Will merge and build later today.

**Pull request merged**